sdk/search/Azure.Search.Documents/src/Batching/AsyncEventExtensions.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Search.Documents.Batching
+{
+    /// <summary>
+    /// Extensions to raise <see cref="Func{T, CancellationToken, Task}"/>
+    /// events that wait for every Task to complete and throw every exception.
+    /// </summary>
+    internal static class AsyncEventExtensions
+    {
+        /// <summary>
+        /// Wait for all tasks to be completed and throw every exception.
+        /// </summary>
+        /// <param name="tasks">The tasks to execute.</param>
+        /// <returns>A Task representing completion of all handlers.</returns>
+        private static async Task JoinAsync(IEnumerable<Task> tasks)
+        {
+            if (tasks != null)
+            {
+                Task joined = Task.WhenAll(tasks);
+                try
+                {
+                    await joined.ConfigureAwait(false);
+                }
+                catch (Exception)
+                {
+                    // awaiting will unwrap the AggregateException which we
+                    // don't want if there were multiple failures that should
+                    // be surfaced
+                    if (joined.Exception?.InnerExceptions?.Count > 1)
+                    {
+                        throw joined.Exception;
```

Comment:
Good point.

Reply:
Looking at this again, I remembered why I didn't care.  This is the AggregateException created and thrown by awaiting a few lines up.  I don't mind shifting its callstack down here because all of its InnerExceptions with the user's callstacks are fine.